### PR TITLE
Check minor version mismatch between client and server before storage migration

### PR DESCRIPTION
**Pull request by nak3**
If `oc adm migrate storage` was ran by different client version from
server's, resource migration could cause errors or be done
unexpectedly.

This patch introduces minor version check betfore running `oc adm
migrate storage`.

### CODE REVIEW

File: pkg/oc/admin/migrate/storage/storage.go [hunk 1]
```diff
@@ -198,6 +208,8 @@ func NewCmdMigrateAPIStorage(name, fullName string, f *clientcmd.Factory, in io.
 	flags.IntVar(&options.bandwidth, "bandwidth", 10,
 		"Average network bandwidth measured in megabits per second (Mbps) to use during storage migration.  Zero means no limit.  This flag is alpha and may change in the future.")
 
+	flags.BoolVar(&options.skipVersionCheck, "skip-versioncheck", false, "Skip minor version check between client and server.")
```
Comment (soltysh) [inline pkg/oc/admin/migrate/storage/storage.go]: @deads2k  I think we don't want to do this skip at all. The migration always needs to be performed using the exact version, or same and higher?

**Comment by enj** [comment]
This is simply incorrect.  The migration of objects is always done by the server. `oc` sends unmodified bytes.  It was *always* safe, regardless of version skew.

**Comment by nak3** [comment]
you can see the issue if you ran oc v3.7 against 3.9 cluster
https://bugzilla.redhat.com/show_bug.cgi?id=1587830

**Comment by enj** [comment]
> you can see the issue if you ran oc v3.9 against 3.7 cluster, https://bugzilla.redhat.com/show_bug.cgi?id=1587830

This is *not* an issue.  The migration command is failing on discovery.  It will *never* cause any harm if it actually manages to run.  Version skew checks are not valid.

**Comment by nak3** [comment]
@enj Based on your explanation, the discovery which command was failing is not supposed to be necessary in migration. Can you remove the discovery code from the migration? (The discovery is called in migration but failing is no harm. It is contradictory.)
Then, migration can be done by any client version and it should be safe according to your explanation.

**Comment by enj** [comment]
> Based on your explanation, the discovery which command was failing is not supposed to be necessary in migration.

It is required to make sure the command attempts to migrate all resources and thus cannot be removed.

> Can you remove the discovery code from the migration?

No per above.

> (The discovery is called in migration but failing is no harm. It is contradictory.)

Running an incomplete migration by skipping discovery is pointless because it gives the false appearance of doing a full migration.  So discovery failing points out that the command cannot fully accomplish its job.  The reason it is safe is because object migration/mutation occurs on the server.  The client determines the list of resources to migrate, which requires discovery.  Not doing discovery would violate an implicit contract of the command.

> Then, migration can be done by any client version 

This is not a goal of the command or the CLI in general.

> it should be safe according to your explanation.

Safe but possibly incomplete, per above.

---

I would not worry too much about this.  Long term we need to work with upstream to move all of this as an automatic server side operation.  It causes way too much pain during upgrades due to long running time and it fails due to random server bugs which it has no control over.

**Pull request closed without merging**